cgen: appending wrong array type to sumtype array generates a cgen error

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
```
OS: macos, macOS, 12.1, 21C52
Processor: 8 cpus, 64bit, little endian, Apple M1
CC version: Apple clang version 13.0.0 (clang-1300.0.29.30)

getwd: /Users/hunam/Documents/dev/go2v
vmodules: /Users/hunam/.vmodules
vroot: /Applications/v
vexe: /Applications/v/v
vexe mtime: 2022-02-23 17:36:00
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 36c9ed6.9662b79

Git version: git version 2.32.0 (Apple Git-132)
Git vroot status: weekly.2022.07-77-g9662b796
.git/config present: true
thirdparty/tcc status: thirdparty-unknown-unknown de82a130
```

<!-- Please include the standalone minimal reproducible code.
If not, it may be closed by the administrators. -->
**What did you do?**
```v
type Bar = BarA | BarB

struct BarA {}

struct BarB {}

fn main() {
	mut bar := []Bar{}
	bar << ['hey']
}
```

**What did you expect to see?**
```v
report.v:9:9: error: cannot append `[]string` to `[]Bar`
    7 | fn main() {
    8 |     mut bar := []Bar{}
    9 |     bar << ['hey']
      |            ~~~~~~~
   10 | }
```
 
**What did you see instead?**
```
/tmp/v_501/report.6964987654114301997.tmp.c:2064:4: warning: expression result unused [-Wunused-value]
  (*(int*)_t1.data);
   ^~~~~~~~~~~~~~~
/tmp/v_501/report.6964987654114301997.tmp.c:11254:88: error: initializing 'main__BarA *' (aka 'struct main__BarA *') with an expression of incompatible type 'string' (aka 'struct string')
        _PUSH_MANY(&bar, (new_array_from_c_array(1, 1, sizeof(main__Bar), _MOV((main__Bar[1]){_SLIT("hey")}))), _t1, Array_main__Bar);
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/report.6964987654114301997.tmp.c:161:18: note: expanded from macro '_SLIT'
#define _SLIT(s) ((string){.str=(byteptr)("" s), .len=(sizeof(s)-1), .is_lit=1})
                 ^
/tmp/v_501/report.6964987654114301997.tmp.c:171:60: note: expanded from macro '_PUSH_MANY'
#define _PUSH_MANY(arr, val, tmp, tmp_typ) {tmp_typ tmp = (val); array_push_many(arr, tmp.data, tmp.len);}
                                                           ^~~
1 warning and 1 error generated.
```

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -845,6 +845,22 @@ pub fn (mut c Checker) infix_expr(mut node ast.InfixExpr) ast.Type {
 							right_pos)
 					}
 					return ast.void_type
+				} else if left_value_sym.kind == .sum_type {
+					if right_final.kind != .array {
+						if left_value_type.idx() != right_type.idx()
+							&& !c.table.sumtype_has_variant(left_value_type, right_type, false) {
+							c.error('cannot append `$right_sym.name` to `$left_sym.name`',
+								right_pos)
+						}
+					} else {
+						right_value_type := c.table.value_type(right_type)
+						if left_value_type.idx() != right_value_type.idx()
+							&& !c.table.sumtype_has_variant(left_value_type, right_value_type, false) {
+							c.error('cannot append `$right_sym.name` to `$left_sym.name`',
+								right_pos)
+						}
+					}
+					return ast.void_type
 				}
 				// []T << T or []T << []T
 				unwrapped_right_type := c.unwrap_generic(right_type)

--- a/vlib/v/checker/containers.v
+++ b/vlib/v/checker/containers.v
@@ -110,7 +110,17 @@ pub fn (mut c Checker) array_init(mut node ast.ArrayInit) ast.Type {
 				continue
 			} else if expecting_sumtype_array {
 				if i == 0 {
-					elem_type = expected_value_type
+					if expected_value_type.idx() == typ.idx()
+						|| c.table.sumtype_has_variant(expected_value_type, typ, false) {
+						elem_type = expected_value_type
+					} else {
+						if expr.is_auto_deref_var() {
+							elem_type = ast.mktyp(typ.deref())
+						} else {
+							elem_type = ast.mktyp(typ)
+						}
+					}
+					c.expected_type = elem_type
 				}
 				continue
 			}

--- a/vlib/v/checker/tests/array_of_sumtype_append_err.out
+++ b/vlib/v/checker/tests/array_of_sumtype_append_err.out
@@ -1,0 +1,13 @@
+vlib/v/checker/tests/array_of_sumtype_append_err.vv:9:9: error: cannot append `[]string` to `[]Bar`
+    7 | fn main() {
+    8 |     mut bar := []Bar{}
+    9 |     bar << ['hey']
+      |            ~~~~~~~
+   10 |     bar << 'hey'
+   11 | }
+vlib/v/checker/tests/array_of_sumtype_append_err.vv:10:9: error: cannot append `string` to `[]Bar`
+    8 |     mut bar := []Bar{}
+    9 |     bar << ['hey']
+   10 |     bar << 'hey'
+      |            ~~~~~
+   11 | }

--- a/vlib/v/checker/tests/array_of_sumtype_append_err.vv
+++ b/vlib/v/checker/tests/array_of_sumtype_append_err.vv
@@ -1,0 +1,11 @@
+type Bar = BarA | BarB
+
+struct BarA {}
+
+struct BarB {}
+
+fn main() {
+	mut bar := []Bar{}
+	bar << ['hey']
+	bar << 'hey'
+}

